crates/matrix-sdk-indexeddb/src/state_store.rs

```diff
@@ -137,137 +168,310 @@ mod KEYS {
-                db.create_object_store(KEYS::DISPLAY_NAMES)?;
-                db.create_object_store(KEYS::JOINED_USER_IDS)?;
-                db.create_object_store(KEYS::INVITED_USER_IDS)?;
-
-                db.create_object_store(KEYS::ROOM_STATE)?;
-                db.create_object_store(KEYS::ROOM_INFOS)?;
-                db.create_object_store(KEYS::PRESENCE)?;
-                db.create_object_store(KEYS::ROOM_ACCOUNT_DATA)?;
-
-                db.create_object_store(KEYS::STRIPPED_ROOM_INFOS)?;
-                db.create_object_store(KEYS::STRIPPED_MEMBERS)?;
-                db.create_object_store(KEYS::STRIPPED_ROOM_STATE)?;
-                db.create_object_store(KEYS::STRIPPED_JOINED_USER_IDS)?;
-                db.create_object_store(KEYS::STRIPPED_INVITED_USER_IDS)?;
-
-                db.create_object_store(KEYS::ROOM_USER_RECEIPTS)?;
-                db.create_object_store(KEYS::ROOM_EVENT_RECEIPTS)?;
-
-                #[cfg(feature = "experimental-timeline")]
-                {
-                    db.create_object_store(KEYS::ROOM_TIMELINE)?;
-                    db.create_object_store(KEYS::ROOM_TIMELINE_METADATA)?;
-                    db.create_object_store(KEYS::ROOM_EVENT_ID_TO_POSITION)?;
-                }
+    for name in ALL_STORES {
+        let tx = target.transaction_on_one_with_mode(name, IdbTransactionMode::Readwrite)?;
 
-                db.create_object_store(KEYS::MEDIA)?;
+        let obj = tx.object_store(name)?;
 
-                db.create_object_store(KEYS::CUSTOM)?;
-            }
-            Ok(())
-        }));
+        if let Some(curs) = source
+            .transaction_on_one_with_mode(name, IdbTransactionMode::Readonly)?
+            .object_store(name)?
+            .open_cursor()?
+            .await?
+        {
+            while let Some(key) = curs.key() {
+                obj.put_key_val(&key, &curs.value())?;
 
-        let db: IdbDatabase = db_req.into_future().await?;
+                curs.continue_cursor()?.await?;
+            }
+        }
 
-        Ok(Self { name, inner: db, store_cipher })
+        tx.await.into_result()?;
     }
 
-    #[allow(dead_code)]
-    pub async fn open() -> StoreResult<Self> {
-        Ok(IndexeddbStore::open_helper("state".to_owned(), None).await?)
-    }
+    let tx =
+        meta.transaction_on_one_with_mode(KEYS::BACKUPS_META, IdbTransactionMode::Readwrite)?;
+    let backup_store = tx.object_store(KEYS::BACKUPS_META)?;
+    backup_store.put_key_val(&JsValue::from_f64(now), &JsValue::from_str(&backup_name))?;
 
-    pub async fn open_with_passphrase(name: String, passphrase: &str) -> StoreResult<Self> {
-        Ok(Self::inner_open_with_passphrase(name, passphrase).await?)
-    }
+    tx.await;
 
-    pub(crate) async fn inner_open_with_passphrase(name: String, passphrase: &str) -> Result<Self> {
-        let name = format!("{:0}::matrix-sdk-state", name);
+    Ok(())
+}
+
+#[derive(Builder, Debug, PartialEq)]
```

Comment:
Should we really be using `derive_builder` for a four-parameter thing? I kind of feel like that just makes it much easier to accidentally have less docs, while hardly making things simpler. This builder doesn't have any useful docs on its methods as-is, right?

Reply:
[`derive_builder` has docs support](https://docs.rs/derive_builder/latest/derive_builder/#documentation-comments-and-attributes) and the compiler is complaining about them being missing.

![grafik](https://user-images.githubusercontent.com/40496/181210943-8f153df9-1b86-47f3-8006-f5027b43f2c7.png)


Another reason I opted for using the derive_builder (other than we have that dependency already anyways), was to extending the configuration builder later a lot easier. Sure, it is only three params now, but it was only two before and through this we can very easily extend things without bloating our code a lot or breaking downstream compatibility.

Reply:
Alright, I guess it's not too bad.